### PR TITLE
Webspace copy command

### PR DESCRIPTION
**Pull request by trickreich**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | --
| Related issues/PRs | --
| License | MIT
| Documentation PR |--

#### What's in this PR?

Implemented copy webspace command.

### CODE REVIEW

File: src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php [hunk 1]
```diff
@@ -0,0 +1,667 @@
...
+     * @param BasePageDocument $documentSource
+     * @param BasePageDocument|null $parentDocumentDestination
+     * @param string $localeDestination
+     */
+    protected function recursiveCopy(
+        BasePageDocument $documentSource,
+        BasePageDocument $parentDocumentDestination = null,
+        $localeDestination
+    ) {
+        // Generate new path.
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php]: Superfluous comment, the variable describes that well enough. Would remove it.

File: src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php [hunk 2]
```diff
@@ -0,0 +1,667 @@
...
+            $this->sessionManager->getContentPath($this->webspaceKeyDestination),
+            $documentSource->getPath()
+        );
+
+        $this->output->writeln('<info>Processing: </info>' . $documentSource->getPath() . ' => ' . $newPath);
+
+        /** @var BasePageDocument $documentDestination */
+        try {
+            $documentDestination = $this->documentManager->find($newPath, $localeDestination);
+        } catch (\Exception $exception) {
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php]: Isn't there a better exception you can catch? What about the `DocumentNotFoundException`?

File: src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php [hunk 3]
```diff
@@ -0,0 +1,667 @@
...
+        $documentDestination->setTitle($documentSource->getTitle());
+        $documentDestination->setStructureType($documentSource->getStructureType());
+        $documentDestination->setWorkflowStage(WorkflowStage::TEST);
+        $documentDestination->setExtensionsData($documentSource->getExtensionsData());
+        $documentDestination->setResourceSegment($documentSource->getResourceSegment());
+        $documentDestination->setPermissions($documentSource->getPermissions());
+        $documentDestination->setSuluOrder($documentSource->getSuluOrder());
+        $documentDestination->setNavigationContexts($documentSource->getNavigationContexts());
+        $documentDestination->setShadowLocaleEnabled($documentSource->isShadowLocaleEnabled());
+        $documentDestination->setShadowLocale($documentSource->getShadowLocale());
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php]: Just an idea: Would cloning the document work? Then we could only call the setters for the properties that really change...

File: src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php [hunk 4]
```diff
@@ -0,0 +1,667 @@
...
+
+    /**
+     * @param BasePageDocument $documentSource
+     * @param string $localeDestination
+     */
+    protected function recursiveCopyRedirectsAndStructure(
+        BasePageDocument $documentSource,
+        $localeDestination
+    ) {
+        // Generate new path.
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php]: Again, superfluous comment

File: src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php [hunk 5]
```diff
@@ -0,0 +1,667 @@
...
+
+        $this->output->writeln('<info>Processing: </info>' . $documentSource->getPath() . ' => ' . $newPath);
+
+        /** @var PageDocument $documentDestination */
+        $documentDestination = $this->documentManager->find($newPath, $localeDestination);
+
+        // Copy the redirects and correct the target.
+        switch ($documentSource->getRedirectType()) {
+            case RedirectType::INTERNAL:
+                // Generate new target path.
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php]: Superfluous comment

File: src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php [hunk 6]
```diff
@@ -0,0 +1,667 @@
...
+                $documentDestination->setRedirectType(RedirectType::INTERNAL);
+                $documentDestination->setRedirectTarget(
+                    $this->documentManager->find($newPathTarget, $localeDestination)
+                );
+                break;
+            case RedirectType::EXTERNAL:
+                $documentDestination->setRedirectType(RedirectType::EXTERNAL);
+                $documentDestination->setRedirectExternal($documentDestination->getRedirectExternal());
+                break;
+            default:
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php]: That's pretty unnecessary, isn't it? 😉

File: src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php [hunk 7]
```diff
@@ -0,0 +1,667 @@
...
+        $localeDestination
+    ) {
+        /** @var BasePageDocument $targetDocumentSource */
+        $targetDocumentSource = $this->documentManager->find($uuid, $localeSource);
+
+        if ($this->webspaceKeySource !== $targetDocumentSource->getWebspaceName()) {
+            return null;
+        }
+
+        // Generate new target path.
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php]: superfluous comment

**Comment by danrot** [comment]
And tests should be working 😉 Once you've fixed that I will test it.

File: src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php [hunk 8]
```diff
@@ -0,0 +1,667 @@
...
+            'Locales: ' . implode(', ', $localesPairs),
+            '---------------',
+            '',
+        ]);
+
+        $helper = $this->getHelper('question');
+        $question = new ConfirmationQuestion('<question>Continue with this options?(y/n)</question> ', false);
+
+        if (!$helper->ask($input, $output, $question)) {
+            $output->writeln('<error>Abort!</error>');
```
Comment (trickreich) [inline src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php]: I've copied this from another command.. in my opinion you are right.. it's no error.. but an info that a user has aborted the command is legit..

File: src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php [hunk 9]
```diff
@@ -0,0 +1,667 @@
...
+     */
+    protected function clearDestinationWebspace()
+    {
+        $homeDocument = $this->documentManager->find(
+            $this->sessionManager->getContentPath($this->webspaceKeyDestination)
+        );
+        foreach ($homeDocument->getChildren() as $child) {
+            $this->output->writeln('<info>Processing: </info>' . $child->getPath());
+            $this->documentManager->remove($child);
+            $this->documentManager->flush();
```
Comment (trickreich) [inline src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php]: Sadly not.. I've got errors when not flushing after every document..

File: src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php (hunk 2 above)
Comment (trickreich) [inline src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php]: yes, will use the correct exception

File: src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php (hunk 3 above)
Comment (trickreich) [inline src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php]: In my opinion not, we just get the data what we really need..

File: src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php (hunk 6 above)
Comment (trickreich) [inline src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php]: I've removed it :)

File: src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php [hunk 10]
```diff
@@ -0,0 +1,675 @@
...
+
+        $output->writeln([
+            '<info>==============================</info>',
+            '<info>ATTENTION</info>',
+            '<info>The whole destination webspace (' . $webspaceKeyDestination . ') will be deleted!</info>',
+            '<info>==============================</info>',
+            '',
+        ]);
+
+        if (!true === $input->getOption('clear-destination-webspace')) {
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php]: `true !== $input->...`

File: src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php [hunk 11]
```diff
@@ -0,0 +1,675 @@
...
+        $output->writeln([
+            '<info>==============================</info>',
+            '<info>ATTENTION</info>',
+            '<info>The whole destination webspace (' . $webspaceKeyDestination . ') will be deleted!</info>',
+            '<info>==============================</info>',
+            '',
+        ]);
+
+        if (true !== $input->getOption('clear-destination-webspace')) {
+            $output->writeln([
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php]: This were the methods I've been talking about: http://symfony.com/doc/current/console/style.html#result-methods

Then the style would be nicer and more consistent.

File: src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php [hunk 12]
```diff
@@ -0,0 +1,675 @@
...
+                '<error>==============================',
+                '<error>Aborted!</error>',
+                '<error>Run this command with --clear-destination-webspace</error>',
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Command/WebspaceCopyCommand.php]: I would add more explanation here: 

```
This command currently does not work if there is already data in the webspace. You can run the command with --clear-destination-webspace to remove all the content from the destination webspace.
```

**Comment by danrot** [comment]
Tests are still failing... And who is using this command afterwards? I'd like him to test it with the real data. If there is still a bug with their data we have to working on fixing it again...

**Comment by danrot** [comment]
There seems to be an error with ghost pages:

image

**Comment by danrot** [comment]
Can you maybe explain what has been fixed in these two commits? You could actually also write that in the commit message 😉

**Comment by trickreich** [comment]
Check always if the searched target exists

**Pull request merged**